File: src/data/tutorials/048-pacific-atlantic-water-flow.js
/**
 * Tutorial content for #048 — Pacific Atlantic Water Flow
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an m×n matrix of heights, find all cells from which rainwater can flow to BOTH the Pacific Ocean (top and left borders) AND the Atlantic Ocean (bottom and right borders). Water flows from a cell to an adjacent cell (up/down/left/right) only if the adjacent cell's height is less than or equal to the current cell's height.`,
    example: `heights=[[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]\n→ Pacific touches top row and left column.\n→ Atlantic touches bottom row and right column.\n→ Cells that can reach both:\n✅ Answer: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]`,
    keyInsight: `Instead of asking "can water from cell X reach both oceans?" (expensive DFS per cell), reverse the question: start BFS/DFS from the ocean borders and find all cells that can REACH each border (water flowing uphill in reverse). Cells reachable from both borders are the answer.`,
  },

  approaches: {
    'Brute Force DFS from Each Cell': {
      intuition: `For every cell in the grid, run a DFS to check if water can flow to the Pacific, and another DFS to check if it can reach the Atlantic. A cell makes it into the answer only if both DFS calls return true. This is straightforward but very slow for large grids.`,
      steps: [
        `For each cell (r, c), run dfsCanReach(r, c, "pacific") and dfsCanReach(r, c, "atlantic").`,
        `dfsCanReach explores downhill/equal neighbours recursively. Pacific is reachable if row=0 or col=0 is hit. Atlantic if row=m-1 or col=n-1.`,
        `Use a visited set per DFS call to avoid infinite loops.`,
        `If both return true, add (r, c) to the result list.`,
      ],
      example: `heights=[[1,2],[3,4]], m=2, n=2\n\nCell (0,0)=1:\n  Pacific? row=0 → yes immediately.\n  Atlantic? From (0,0) go to (1,0)=3 or (0,1)=2.\n    From (1,0) go to (1,1)=4 → col=n-1=1 and row=m-1=1 → Atlantic reached!\n  Both reached → add (0,0).\nCell (0,1)=2: Pacific? row=0 → yes. Atlantic? (1,1)=4 → yes. Add (0,1).\nCell (1,0)=3: Pacific? → (0,0)=1<3 can't go, (1,0) row=1≠0, col=0 → Pacific reached! Atlantic? row=1=m-1 → yes. Add (1,0).\nCell (1,1)=4: Both borders trivially reachable. Add (1,1).\n✅ Answer: [[0,0],[0,1],[1,0],[1,1]]`,
      keyInsight: `O(m² × n²) time — a DFS for every cell. Very slow for large grids. Illustrates the problem clearly but is not viable for m,n > ~50.`,
    },

    'Multi-Source BFS (Optimal)': {
      intuition: `Start BFS from ALL Pacific border cells simultaneously, flowing "uphill" (to cells with equal or greater height). Mark every cell reachable as "pacific-reachable". Do the same from ALL Atlantic border cells. Cells marked in both sets are the answer. This visits each cell at most twice — optimal.`,
      steps: [
        `Create two visited sets: pacificReachable and atlanticReachable.`,
        `Seed the Pacific BFS queue with all cells in row 0 and column 0.`,
        `Seed the Atlantic BFS queue with all cells in row m-1 and column n-1.`,
        `BFS rule: from cell (r,c), move to neighbour (nr,nc) if heights[nr][nc] >= heights[r][c] (water flows uphill in reverse).`,
        `Run both BFS passes to completion.`,
        `Collect all cells (r,c) present in both reachable sets.`,
      ],
      example: `heights=[[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]\n\nPacific BFS seeds: entire top row + left column.\n  From (0,4)=5: can flow to (1,4)=4? 4<5, no. Stays.\n  From (4,0)=5: reaches (3,0)=6≥5 → add (3,0). From (3,0)=6: (2,0)=2<6 stop. (3,1)=7≥6 → add (3,1).\nAtlantic BFS seeds: bottom row + right column.\n  From (0,4)=5: Atlantic border directly.\n  From (4,0)=5: Atlantic border directly.\n\nIntersection of both sets:\n✅ Answer: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]`,
      keyInsight: `O(m×n) time and space. Multi-source BFS is the key insight — treat ALL border cells as simultaneous starting points. Reversing the flow direction converts "can water drain to ocean?" into "can ocean expand inward?"`,
    },

    'Multi-Source DFS (Optimal DFS variant)': {
      intuition: `Same multi-source "reverse flow" insight as BFS, but use DFS instead of a queue. Start DFS simultaneously from all Pacific border cells (top row + left column), marking every cell reachable going uphill. Then start DFS from all Atlantic border cells (bottom row + right column). A cell is in the answer if it was marked by both DFS passes.`,
      steps: [
        `Create two boolean grids: pacific[m][n] and atlantic[m][n], both false.`,
        `Run dfs(r, c, reachable) from every cell in row 0 and column 0, marking pacific reachable.`,
        `Run dfs(r, c, reachable) from every cell in row m-1 and column n-1, marking atlantic reachable.`,
        `DFS rule: from (r,c), move to neighbour (nr,nc) if heights[nr][nc] >= heights[r][c] AND not yet marked.`,
        `Collect all (r,c) where pacific[r][c] && atlantic[r][c] is true.`,
      ],
      example: `heights=[[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]\n\nPacific DFS seeds: row 0 + col 0. Mark all cells reachable flowing uphill from Pacific border.\nAtlantic DFS seeds: row 4 + col 4. Mark all cells reachable flowing uphill from Atlantic border.\n\nCells marked in both:\n(0,4): pacific=true (row 0 border), atlantic=true (col 4 border) → both ✓\n(1,3): pacific reachable via uphill path from top, atlantic reachable via uphill path from right border → both ✓\n... (similar for remaining answer cells)\n✅ Answer: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]`,
      keyInsight: `O(m×n) time, O(m×n) space. Identical complexity to Multi-Source BFS. DFS uses the call stack (O(m×n) worst case) rather than a queue. Use BFS if very large grids risk stack overflow; use DFS for simpler recursive code.`,
    },
  },
}

File: public/cpp/048-pacific-atlantic-water-flow.cpp
/*
 * LeetCode Problem #417: Pacific Atlantic Water Flow
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/pacific-atlantic-water-flow/
 *
 * Problem Statement:
 * There is an m x n rectangular island that borders both the Pacific Ocean (top/left)
 * and the Atlantic Ocean (bottom/right). Rain water can flow to a neighboring cell
 * if it is at the same or lower height. Return all cells from which water can flow
 * to BOTH oceans.
 *
 * Example:
 *   heights = [[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]
 *   Output: [[0,4],[1,3],[1,4],[2,2],[3,0],[3,1],[4,0]]
 *
 * Constraints:
 *   - m == heights.length, n == heights[r].length
 *   - 1 <= m, n <= 200
 *   - 0 <= heights[i][j] <= 10^5
 */

#include <vector>
#include <queue>
#include <iostream>
using namespace std;

using vvi = vector<vector<int>>;
using vvb = vector<vector<bool>>;

// ==================== APPROACH 1: Brute Force (DFS from each cell) ====================
// For each cell, DFS to see if it can reach Pacific and Atlantic independently.
// Time: O(m*n*(m+n))  Space: O(m*n)
class Solution1 {
    int m, n;
    bool canReach(int r, int c, const vvi& h, bool pacific) {
        vector<vector<bool>> vis(m, vector<bool>(n, false));
        return dfs(r, c, h, vis, pacific);
    }
    bool dfs(int r, int c, const vvi& h, vvb& vis, bool pacific) {
        if (pacific && (r == 0 || c == 0)) return true;
        if (!pacific && (r == m-1 || c == n-1)) return true;
        vis[r][c] = true;
        int dr[] = {0,0,1,-1}, dc[] = {1,-1,0,0};
        for (int d = 0; d < 4; d++) {
            int nr = r+dr[d], nc = c+dc[d];
            if (nr<0||nr>=m||nc<0||nc>=n||vis[nr][nc]) continue;
            if (h[nr][nc] <= h[r][c] && dfs(nr, nc, h, vis, pacific)) return true;
        }
        return false;
    }
public:
    vvi pacificAtlantic(vvi& heights) {
        m = heights.size(); n = heights[0].size();
        vvi res;
        for (int r=0;r<m;r++) for (int c=0;c<n;c++)
            if (canReach(r,c,heights,true) && canReach(r,c,heights,false))
                res.push_back({r,c});
        return res;
    }
};

// ==================== APPROACH 2: BFS Multi-source (Optimal) ====================
// Reverse-flood from each ocean border (water flows uphill from ocean).
// Cells reachable from both = answer.
// Time: O(m*n)  Space: O(m*n)
class Solution2 {
    void bfs(queue<pair<int,int>>& q, vvb& visited, const vvi& h) {
        int dr[] = {0,0,1,-1}, dc[] = {1,-1,0,0};
        while (!q.empty()) {
            auto [r,c] = q.front(); q.pop();
            for (int d=0;d<4;d++) {
                int nr=r+dr[d], nc=c+dc[d];
                if (nr<0||nr>=(int)h.size()||nc<0||nc>=(int)h[0].size()||visited[nr][nc]) continue;
                if (h[nr][nc] >= h[r][c]) { visited[nr][nc]=true; q.push({nr,nc}); }
            }
        }
    }
public:
    vvi pacificAtlantic(vvi& heights) {
        int m=heights.size(), n=heights[0].size();
        vvb pac(m,vector<bool>(n,false)), atl(m,vector<bool>(n,false));
        queue<pair<int,int>> pq, aq;
        for (int r=0;r<m;r++) { pac[r][0]=true; pq.push({r,0}); atl[r][n-1]=true; aq.push({r,n-1}); }
        for (int c=0;c<n;c++) { pac[0][c]=true; pq.push({0,c}); atl[m-1][c]=true; aq.push({m-1,c}); }
        bfs(pq,pac,heights); bfs(aq,atl,heights);
        vvi res;
        for (int r=0;r<m;r++) for (int c=0;c<n;c++) if (pac[r][c]&&atl[r][c]) res.push_back({r,c});
        return res;
    }
};

// ==================== APPROACH 3: DFS Multi-source (Optimal DFS variant) ====================
// Same reverse-flood idea using DFS.
// Time: O(m*n)  Space: O(m*n)
class Solution3 {
    void dfs(int r, int c, vvb& visited, const vvi& h) {
        visited[r][c] = true;
        int dr[]={0,0,1,-1}, dc[]={1,-1,0,0};
        for (int d=0;d<4;d++) {
            int nr=r+dr[d], nc=c+dc[d];
            if (nr<0||nr>=(int)h.size()||nc<0||nc>=(int)h[0].size()||visited[nr][nc]) continue;
            if (h[nr][nc] >= h[r][c]) dfs(nr,nc,visited,h);
        }
    }
public:
    vvi pacificAtlantic(vvi& heights) {
        int m=heights.size(), n=heights[0].size();
        vvb pac(m,vector<bool>(n,false)), atl(m,vector<bool>(n,false));
        for (int r=0;r<m;r++) { if (!pac[r][0]) dfs(r,0,pac,heights); if (!atl[r][n-1]) dfs(r,n-1,atl,heights); }
        for (int c=0;c<n;c++) { if (!pac[0][c]) dfs(0,c,pac,heights); if (!atl[m-1][c]) dfs(m-1,c,atl,heights); }
        vvi res;
        for (int r=0;r<m;r++) for (int c=0;c<n;c++) if (pac[r][c]&&atl[r][c]) res.push_back({r,c});
        return res;
    }
};

void runTests() {
    Solution2 sol;
    vvi h1={{1,2,2,3,5},{3,2,3,4,4},{2,4,5,3,1},{6,7,1,4,5},{5,1,1,2,4}};
    auto r1 = sol.pacificAtlantic(h1);
    cout << "Test 1 result size: " << r1.size() << " (expected 7)\n";
    vvi h2={{1}};
    auto r2 = sol.pacificAtlantic(h2);
    cout << "Test 2: " << r2.size() << " (expected 1)\n";
}

int main() { runTests(); return 0; }

// Made with Bob
